statistics/handle/handle.go

```diff
@@ -221,6 +221,118 @@ func (h *Handle) Update(is infoschema.InfoSchema) error {
 	return nil
 }
 
+// GlobalStats is used to store the statistics contained in the global-level stats
+// which is generated by the merge of partition-level stats.
+// It will both store the column stats and index stats.
+// In the column statistics, the variable `num` is equal to the number of columns in the partition table.
+// In the index statistics, the variable `num` is always equal to one.
+type GlobalStats struct {
+	Num   int
+	Count int64
+	Hg    []*statistics.Histogram
+	Cms   []*statistics.CMSketch
+	TopN  []*statistics.TopN
+}
+
+// MergePartitionStats2GlobalStats merge the partition-level stats to global-level stats based on the tableID.
+func (h *Handle) MergePartitionStats2GlobalStats(is infoschema.InfoSchema, physicalID int64, isIndex int, idxID int64) (globalStats *GlobalStats, succ bool) {
+	// get the partition table IDs
+	h.mu.Lock()
+	globalTable, ok := h.getTableByPhysicalID(is, physicalID)
+	h.mu.Unlock()
+	if !ok {
+		logutil.BgLogger().Debug("unknown physical ID in stats meta table, maybe it has been dropped", zap.Int64("ID", physicalID))
+		return
+	}
+	globalTableInfo := globalTable.Meta()
+	partitionNum := globalTableInfo.Partition.Num
+	partitionIDs := make([]int64, 0, partitionNum)
+	for i := uint64(0); i < partitionNum; i++ {
+		partitionIDs = append(partitionIDs, globalTableInfo.Partition.Definitions[i].ID)
+	}
+
+	// initialized the globalStats
+	globalStats = new(GlobalStats)
+	if isIndex == 0 {
+		globalStats.Num = len(globalTableInfo.Columns)
+	} else {
+		globalStats.Num = 1
+	}
+	globalStats.Count = 0
+	globalStats.Hg = make([]*statistics.Histogram, globalStats.Num)
+	globalStats.Cms = make([]*statistics.CMSketch, globalStats.Num)
+	globalStats.TopN = make([]*statistics.TopN, globalStats.Num)
+
+	// The first dimension of slice is means the number of stats in the globalStats.
+	// The second dimension of slice is means the number of partition tables.
+	// We should store all of the partition-level stats first, and merge them together.
+	allHg := make([][]*statistics.Histogram, globalStats.Num)
+	allCms := make([][]*statistics.CMSketch, globalStats.Num)
+	allTopN := make([][]*statistics.TopN, globalStats.Num)
+	for i := 0; i < globalStats.Num; i++ {
+		allHg[i] = make([]*statistics.Histogram, 0, partitionNum)
+		allCms[i] = make([]*statistics.CMSketch, 0, partitionNum)
+		allTopN[i] = make([]*statistics.TopN, 0, partitionNum)
+	}
+
+	for _, partitionID := range partitionIDs {
+		h.mu.Lock()
+		partitionTable, ok := h.getTableByPhysicalID(is, partitionID)
+		h.mu.Unlock()
+		if !ok {
+			logutil.BgLogger().Debug("unknown physical ID in stats meta table, maybe it has been dropped", zap.Int64("ID", partitionID))
+			return
+		}
+		tableInfo := partitionTable.Meta()
+		partitionStats, err := h.TableStatsFromStorage(tableInfo, partitionID, false, nil)
```

Comment:
We can put this in the cache. But I think it's a tradeoff. Because we just use the partition-level stats to build the global-level stats. Hardly used in other situations. This will take up a lot of cache memory to store these stats that are likely to be used only once.

Reply:
Partition-level stats can be divided into two kinds. One kind is not involved in current analyze and we just read stats from storage to participate in the merging process. The other kind is newly calculated and saved to storage in     current analyze. Reading the second kind stats from storage seems unnecessary since we have the stats in memory.

Reply:
After discussion with @qw4990 , we think maybe we can do this optimization later. It is a low-frequency operation. What's your opinion @xuyifangreeneyes